core/src/test/scala/unit/kafka/network/ConnectionQuotasTest.scala

```diff
@@ -302,21 +312,295 @@ class ConnectionQuotasTest {
+      verifyNoBlockedPercentRecordedOnAllListeners()
+
+      verifyConnectionCountOnEveryListener(connectionQuotas, connectionsPerListener)
+    } finally {
+      executor.shutdownNow()
+    }
+  }
+
+  @Test
+  def testListenerConnectionRateLimitWhenActualRateAboveLimit(): Unit = {
+    val brokerRateLimit = 125
+    val listenerRateLimit = 30
+    val connCreateIntervalMs = 25 // connection creation rate = 40/sec per listener (3 * 40 = 120/sec total)
+    val props = brokerPropsWithDefaultConnectionLimits
+    props.put(KafkaConfig.MaxConnectionCreationRateProp, brokerRateLimit.toString)
+    val config = KafkaConfig.fromProps(props)
+    val connectionQuotas = new ConnectionQuotas(config, Time.SYSTEM, metrics)
+
+    val listenerConfig = Map(KafkaConfig.MaxConnectionCreationRateProp -> listenerRateLimit.toString).asJava
+    addListenersAndVerify(config, listenerConfig, connectionQuotas)
+
+    val executor = Executors.newFixedThreadPool(listeners.size)
+    try {
+      // create connections with the rate > listener quota on every listener
+      // run a bit longer (20 seconds) to also verify the throttle rate
+      val connectionsPerListener = 600 // should take 20 seconds to create 600 connections with rate = 30/sec
+      val futures = listeners.values.map { listener =>
+        executor.submit((() =>
+          // epsilon is set to account for the worst-case where the measurement is taken just before or after the quota window
+          acceptConnectionsAndVerifyRate(connectionQuotas, listener, connectionsPerListener, connCreateIntervalMs, listenerRateLimit, 5)): Runnable)
+      }
+      futures.foreach(_.get(30, TimeUnit.SECONDS))
+
+      // verify that every listener was throttled
+      blockedPercentMeters.foreach { case (name, meter) =>
+        assertTrue(s"Expected BlockedPercentMeter metric for $name listener to be recorded", meter.count() > 0)
+      }
+
+      // while the connection creation rate was throttled,
+      // expect all connections got created (not limit on the number of connections)
+      verifyConnectionCountOnEveryListener(connectionQuotas, connectionsPerListener)
+    } finally {
+      executor.shutdownNow()
+    }
+  }
+
+  @Test
+  def testMaxListenerConnectionListenerMustBeAboveZero(): Unit = {
+    val config = KafkaConfig.fromProps(brokerPropsWithDefaultConnectionLimits)
+    val connectionQuotas = new ConnectionQuotas(config, Time.SYSTEM, metrics)
+
+    connectionQuotas.addListener(config, listeners("EXTERNAL").listenerName)
+
+    val maxListenerConnectionRate = 0
+    val listenerConfig = Map(KafkaConfig.MaxConnectionCreationRateProp -> maxListenerConnectionRate.toString).asJava
+    assertThrows[ConfigException] {
+      connectionQuotas.maxConnectionsPerListener(listeners("EXTERNAL").listenerName).validateReconfiguration(listenerConfig)
+    }
+  }
+
+  @Test
+  def testMaxListenerConnectionRateReconfiguration(): Unit = {
```

Comment:
I wonder if we could just check that the metric's config is correctly re-configured instead of testing the number of connections accepted. The goal of the test is not really to verify that the quota works but rather to ensure that metric is correctly re-configured. Have you considered this? The same would apply to `testMaxBrokerConnectionRateReconfiguration`.

Reply:
I still think it's important to test all the way that it does throttle to the correct value. We had a bug pretty recently, KAFKA-9658, where adding and then removing bandwidth quotas via a dynamic config changed the config and even changed some data structs, but not all data structs to actually change the throttling behavior. So, I feel more confident keeping the test up to checking the throttle.

Reply:
Fair enough.